src/Data/List/Base.agda

```diff
@@ -292,6 +293,16 @@ span P? (x ∷ xs) with does (P? x)
 break : ∀ {P : Pred A p} → Decidable P → List A → (List A × List A)
 break P? = span (∁? P?)
 
+nub-filter : ∀ {R : Rel A p} → B.Decidable R → A → List A → List A
+nub-filter R? y [] = []
+nub-filter R? y (x ∷ xs) with does (R? y x)
+... | true  = nub-filter R? y xs
+... | false = x ∷ nub-filter R? x xs
+
```

Comment:
I think we should use `filter (¬? ∘ R? y)` rather than implementing the ad-hoc `nub-filter`.
This should also make the proofs simpler: we can rely on what we already know about `filter`!

Reply:
@gallais Those are two different functions:

- `nub [1, 1, 2, 1]` is `[1, 2, 1]`
- `dedup [1, 1, 2, 1]` is `[1, 2]`

I happen to need `nub` for further work (also - I know performance is not agda-stdlib goal - `nub` is O(n) while `dedup` is O(n²).

It's true that:

```
nub-filter y xs with span R? xs
nub-filter y xs | _ , [] = []
nub-filter y xs | _ , (z ∷ zs) = nub-filter z zs
```

But I think that it will be hard to convince termination-checker that `zs < xs`

I will add `dedup` as well but this is different function.

Reply:
Ah yup, it is different. However the Haskell [`nub`](https://hoogle.haskell.org/?hoogle=nub) function seems to remove all duplicates, not just consecutive runs so I'm not sure that your function should really be called `nub`.

In general I kind of dislike the name `nub` as it's really not very descriptive, in the Haskell docs they claim it refers to the "essence" but that doesn't make any more sense to me either. I'd be tempted to name the two functions `deduplicate` and `derun`. Thoughts?

> dedup is O(n³)

As an aside, I'm pretty sure recent work on the implementation of decidable equality has got this down to O(n²).

Reply:
> Ah yup, it is different. However the Haskell nub function seems to remove all duplicates, 

Sorry. My fault - blame my reading comprehention

> In general I kind of dislike the name nub as it's really not very descriptive, in the Haskell docs they claim it refers to the "essence" but that doesn't make any more sense to me either. I'd be tempted to name the two functions deduplicate and derun. Thoughts?

I'd think `deduplication` to be on long side. What about `dedup`? I'll rename `nub` to `derun`.

Reply:
> I'd think deduplication to be on long side.

Looking through `Data.List.Base` it looks like the vast majority of function names are written in full, e.g. `intersperse`, `unalignWith`, `replicate`, `tabulate`, `partition` etc. I think `deduplicate` would be more in keeping with these than `dedup` would be.